### PR TITLE
feat: add allowedPropagatingAPIs option for resource selection

### LINKED ISSUE #633: [feature] Add AllowedAPIs option for the resources propagated from the namespace
```
// SkippedPropagatingAPIs indicates comma separated resources that should be skipped for propagating.
SkippedPropagatingAPIs string
```
All resources in the namespace configured with CRP are propagated to the member clusters by default. The hub agent option [`SkippedPropagatingAPIs`](https://github.com/Azure/fleet/blob/6fd37978c0f695b702a069ca6cf9e96c0d2585dc/cmd/hubagent/options/options.go#L52C1-L52C1) can be used to exclude propagating certain APIs from the namespace. The list of skipped APIs can get long to specify all resources and it needs to be updated when the new use cases come up. Alternatively an `AllowedAPIs` option can allow us to be selective while propagating resources. Having a limited set of allowed resources should also reduce reconciliations on the CRPs as it might not trigger on any change in the namespace.

### PR DESCRIPTION
**Pull request by vasudev-bongale**
### Changes

- Add `--allowed-propagating-apis` option to hub-agent command.
- Validate that `--allowed-propagating-apis` and `--skipped-propagating-apis` are mutually exclusive and cannot be specified together.
- Refactor `DisabledResourceConfig` type into a generic `ResourceConfig` so that we can reuse it for both `Allowed` and `Disabled` resource configs. During initializing `DisabledResourceConfig`, fleet and built-in resources are configured.
- Update `change_detector` to watch only the allowed resources if specified.
- Update `resource_selector` to only select allowed resources if specified.
- Add a util function with unit test to determine if a ResourceConfig is empty

#### Other changes
- Add make `help` recipe for developers to list the available commands in the Makefile.

Fixes #633 

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

Verify that the options are mutually exclusive.
```
E1213 17:04:40.247238   91214 main.go:79] "invalid parameter" err="Options.AllowedPropagatingAPIs: Invalid value: \"policy/v1\": AllowedPropagatingAPIs and SkippedPropagatingAPIs are mutually exclusive"
exit status 1
```

Run hub-agent locally, its able to start with the right option value for `--allowed-propagating-apis`
```
❯ export POD_NAMESPACE=fleet-system
❯ go run ./cmd/hubagent/main.go --enable-v1alpha1-apis=false --enable-v1beta1-apis=true --allowed-propagating-apis=rbac.authorization.k8s.io/v1 --v=1
I1214 09:19:00.292904   85484 main.go:76] "flag:" name="add_dir_header" value="false"
I1214 09:19:00.293211   85484 main.go:76] "flag:" name="allowed-propagating-apis" value="rbac.authorization.k8s.io/v1"
```

<details>
<summary>Verify that the Allowed resources are being watched and informer is created</summary>

```
I1214 09:19:11.823118   85484 change_dector.go:205] "Watching Allowed Resource" GVK="rbac.authorization.k8s.io/v1, Kind=ClusterRole"
I1214 09:19:11.823254   85484 change_dector.go:205] "Watching Allowed Resource" GVK="rbac.authorization.k8s.io/v1, Kind=Role"
I1214 09:19:11.823345   85484 change_dector.go:205] "Watching Allowed Resource" GVK="rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
I1214 09:19:11.823394   85484 change_dector.go:205] "Watching Allowed Resource" GVK="rbac.authorization.k8s.io/v1, Kind=RoleBinding"
I1214 09:19:11.826453   85484 informermanager.go:123] "Added an informer for a new resource" res={"GroupVersionKind":{"Group":"rbac.authorization.k8s.io","Version":"v1","Kind":"ClusterRole"},"GroupVersionResource":{"Group":"rbac.authorization.k8s.io","Version":"v1","Resource":"clusterroles"},"IsClusterScoped":true}
I1214 09:19:11.826502   85484 informermanager.go:123] "Added an informer for a new resource" res={"GroupVersionKind":{"Group":"rbac.authorization.k8s.io","Version":"v1","Kind":"Role"},"GroupVersionResource":{"Group":"rbac.authorization.k8s.io","Version":"v1","Resource":"roles"},"IsClusterScoped":false}
I1214 09:19:11.826519   85484 informermanager.go:123] "Added an informer for a new resource" res={"GroupVersionKind":{"Group":"rbac.authorization.k8s.io","Version":"v1","Kind":"ClusterRoleBinding"},"GroupVersionResource":{"Group":"rbac.authorization.k8s.io","Version":"v1","Resource":"clusterrolebindings"},"IsClusterScoped":true}
I1214 09:19:11.826536   85484 informermanager.go:123] "Added an informer for a new resource" res={"GroupVersionKind":{"Group":"rbac.authorization.k8s.io","Version":"v1","Kind":"RoleBinding"},"GroupVersionResource":{"Group":"rbac.authorization.k8s.io","Version":"v1","Resource":"rolebindings"},"IsClusterScoped":false}
```

</details>

<details>
<summary>It selects the correct resources</summary>

```
    - lastTransitionTime: "2023-12-14T03:54:43Z"
      message: Successfully applied resources
      observedGeneration: 1
      reason: ApplySucceeded
      status: "True"
      type: ResourceApplied
  selectedResources:
  - group: rbac.authorization.k8s.io
    kind: Role
    name: vbongale-test-lease-role
    namespace: vbongale-test-ns
    version: v1
  - kind: Namespace
    name: vbongale-test-ns
    version: v1
```

</details>

```
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^TestResourceConfigIsEmpty$ go.goms.io/fleet/pkg/utils
ok  	go.goms.io/fleet/pkg/utils
```

<details>
<summary>$ make help</summary>

```

❯ make help

Usage:
  make <target>
  help             Display this help.
  lint-full        Run slower linters to detect possible issues
  fmt              Run go fmt against code.
  vet              Run go vet against code.
  test             Run tests.
  local-unit-test  Run tests.
  integration-test  Run tests.
  build            Build agent binaries.
  run-hubagent     Run a controllers from your host.
  run-memberagent  Run a controllers from your host.
  clean-bin        Remove all generated binaries
```

</details>

### CODE REVIEW

File: pkg/utils/apiresources.go [hunk 1]
```diff
@@ -162,9 +175,9 @@ func (r *DisabledResourceConfig) parseSingle(token string) error {
 	return nil
 }
 
-// IsResourceDisabled returns whether a given GroupVersionKind is disabled.
-// a gkv is disabled if its group or group version is disabled
-func (r *DisabledResourceConfig) IsResourceDisabled(gvk schema.GroupVersionKind) bool {
```
Comment (zhiying-lin) [inline pkg/utils/apiresources.go:167]: Should we keep the original interface?

you can have something like, and then you don't need to update the caller code.

```
func (r *ResourceConfig) IsResourceDisabled(gvk schema.GroupVersionKind) bool {
	isConfigured := r.isResourceConfigured(gvk)
	if r.isAllowedList {
		return !isConfigured
	}
	return isConfigured
}
```
and change the "IsResourceConfigured" to "isResourceConfigured", which is more like an internal util func.

File: cmd/hubagent/workload/setup.go [hunk 2]
```diff
@@ -105,12 +105,17 @@ func SetupControllers(ctx context.Context, wg *sync.WaitGroup, mgr ctrl.Manager,
...
 	}
 
-	disabledResourceConfig := utils.NewDisabledResourceConfig()
+	disabledResourceConfig := utils.NewResourceConfig(true)
 	if err := disabledResourceConfig.Parse(opts.SkippedPropagatingAPIs); err != nil {
 		// The program will never go here because the parameters have been checked
 		return err
 	}
 
+	allowedResourceConfig := utils.NewResourceConfig(false)
```
Comment (zhiying-lin) [inline cmd/hubagent/workload/setup.go:114]: only one of them can be set. You don't need to maintain two configurations. Please see the comments below.

File: pkg/controllers/clusterresourceplacement/resource_selector.go [hunk 3]
```diff
@@ -287,18 +287,29 @@ func (r *Reconciler) fetchAllResourcesInOneNamespace(namespaceName string, place
 	return resources, nil
 }
 
-// shouldSelectResource returns whether a resource should be propagated
+// shouldSelectResource returns whether a resource should be selected for propagation
```
Comment (zhiying-lin) [inline pkg/controllers/clusterresourceplacement/resource_selector.go:290]:
```suggestion
// shouldSelectResource returns whether a resource should be selected for propagation.
```
missing period

File: pkg/utils/apiresources.go [hunk 4]
```diff
@@ -44,41 +44,54 @@ var (
 	apiGroupSepToken = ";"
 )
 
-// DisabledResourceConfig represents the configuration that identifies the API resources should not be selected.
-type DisabledResourceConfig struct {
-	// groups holds a collection of API group, all resources under this group will be avoided.
+// ResourceConfig represents the configuration that identifies the API resources that are parsed from the
+// user input to either allow or disable propagating them.
+type ResourceConfig struct {
```
[the remark this answers is not in the document]
Comment (vasudev-bongale) [reply pkg/utils/apiresources.go:49]: Having a single Config makes sense since the callers need not change. Thanks for the suggestion. I still renamed `DisabledResourceConfig` to `ResourceConfig` so that we can use for both with a flag.

File: pkg/controllers/clusterresourceplacement/placement_controllerv1alpha1.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ type Reconciler struct {
 	UncachedReader client.Reader
 
 	// DisabledResourceConfig contains all the api resources that we won't select.
-	DisabledResourceConfig *utils.DisabledResourceConfig
+	DisabledResourceConfig *utils.ResourceConfig
```
Comment (ryanzhang-oss) [inline pkg/controllers/clusterresourceplacement/placement_controllerv1alpha1.go:53]:
```suggestion
	ResourceConfig *utils.ResourceConfig
```

File: pkg/resourcewatcher/change_dector.go [hunk 6]
```diff
@@ -65,7 +65,7 @@ type ChangeDetector struct {
 	InformerManager informer.Manager
 
 	// DisabledResourceConfig contains all the api resources that we won't select
-	DisabledResourceConfig *utils.DisabledResourceConfig
+	DisabledResourceConfig *utils.ResourceConfig
```
Comment (ryanzhang-oss) [inline pkg/resourcewatcher/change_dector.go:68]:
```suggestion
	ResourceConfig *utils.ResourceConfig
```

File: pkg/utils/apiresources_test.go [hunk 7]
```diff
@@ -305,6 +305,24 @@ func TestDisabledResourceConfigMixedParse(t *testing.T) {
 			}
 		}
 	}
+	for _, test := range tests {
+		// test allow list
+		r := NewResourceConfig(true)
+		if err := r.Parse(test.input); err != nil {
+			t.Fatalf("Unexpected error: %v", err)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:312]:
```suggestion
			t.Fatalf("Parse() returned error: %v", err)
```

https://go.dev/wiki/TestComments#identify-the-function
`In most tests, failure messages should include the name of the function that failed, even though it seems obvious from the name of the test function.`

File: pkg/controllers/clusterresourceplacement/resource_selector.go [hunk 8]
```diff
@@ -287,9 +287,9 @@ func (r *Reconciler) fetchAllResourcesInOneNamespace(namespaceName string, place
 	return resources, nil
 }
 
-// shouldSelectResource returns whether a resource should be propagated
+// shouldSelectResource returns whether a resource should be selected for propagation.
 func (r *Reconciler) shouldSelectResource(gvr schema.GroupVersionResource) bool {
-	if r.DisabledResourceConfig == nil {
+	if r.ResourceConfig.IsEmpty() {
```
[the remark this answers is not in the document]
Comment (vasudev-bongale) [reply pkg/controllers/clusterresourceplacement/resource_selector.go:292]: `What if the isAllow is true and resourceConfig is empty? we should not select the resource.` - Agreed.

However the ResourceConfig will never be a nil, instead an empty struct is returned by the following function, right? Hence I added another method to check if the ResourceConfig is actually empty of any resources in it. 
```
func NewResourceConfig(isAllowList bool) *ResourceConfig {
	r := &ResourceConfig{
		groups:            map[string]struct{}{},
		groupVersions:     map[schema.GroupVersion]struct{}{},
		groupVersionKinds: map[schema.GroupVersionKind]struct{}{},
	}
```

File: pkg/utils/apiresources_test.go [hunk 9]
```diff
@@ -288,7 +288,7 @@ func TestDisabledResourceConfigMixedParse(t *testing.T) {
 		},
 	}
 	for _, test := range tests {
-		r := NewDisabledResourceConfig()
+		r := NewResourceConfig(false)
```
[the remark this answers is not in the document]
Comment (vasudev-bongale) [reply pkg/utils/apiresources_test.go]: I refactored both `TestDisabledResourceConfigMixedParse` and the DefaultResourceConfig tests.

File: pkg/controllers/clusterresourceplacement/resource_selector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/clusterresourceplacement/resource_selector.go:292]: To make it clear, there are two issues.

1. about "nil", right now, we always initialize the reconciler using NewResourceConfig(), but you cannot prevent other caller creating a reconciler without setting the resourceConfig. It's better to add the check to prevent the regression/panic in the future.
2. about adding r.ResourceConfig.IsEmpty() as a shortcut, i'm fine with checking whether it's empty or not as a shortcut though it's redundant and existing code should cover that.

If you want to add that, it should be

1) expose the isAllowList value
2) and then based on this value to return,

eg, if r.ResourceConfig.IsEmpty() {
      return !r.ResourceConfig.isAllowList
}

Could you please add a UT to test this func to prevent the regression? today we have the tests covered for the SkippedPropagatingAPIs list, but not for the AllowedPropagatingAPIs.

File: pkg/utils/apiresources.go [hunk 10]
```diff
@@ -180,17 +198,22 @@ func (r *DisabledResourceConfig) IsResourceDisabled(gvk schema.GroupVersionKind)
 	return false
 }
 
-// DisableGroup to disable group.
-func (r *DisabledResourceConfig) DisableGroup(g string) {
+// AddGroup to store group in the resource config.
```
Comment (zhiying-lin) [inline pkg/utils/apiresources.go:201]:
```suggestion
// AddGroup stores group in the resource config.
```

Could you please help us to fix the comments? The existing codes are not following the golang best practices.  :(

` Comments documenting declarations should be full sentences, even if that seems a little redundant. This approach makes them format well when extracted into godoc documentation. Comments should begin with the name of the thing being described and end in a period:`
https://go.dev/wiki/CodeReviewComments#comment-sentences

File: pkg/utils/apiresources_test.go [hunk 11]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			r := NewResourceConfig(test.isAllowList)
+			if err := r.Parse(input); err != nil {
+				t.Fatalf("Parse() returned error: %v", err)
+			}
+
+			for i, o := range test.disabled {
+				if ok := r.IsResourceDisabled(o); !ok {
+					t.Errorf("%d: expected resource to be disabled : %v", i, o)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:312]:
```suggestion
					t.Errorf("IsResourceDisabled(%v) = false, want true", o)
```
Could you please fix the test code to follow the test output format?

we're following golang best practices here
`Test outputs should output the actual value that the function returned before printing the value that was expected. A usual format for printing test outputs is “YourFunc(%v) = %v, want %v”.`
https://go.dev/wiki/TestComments#got-before-want

`Do not use the index of the test in the test table as a substitute for naming your tests or printing the inputs. Nobody wants to go through your test table and count the entries in order to figure out which test case is failing.`
https://go.dev/wiki/TestComments#identify-the-input

File: pkg/utils/apiresources_test.go [hunk 12]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+				}
+			}
+
+			for i, o := range test.enabled {
+				if ok := r.IsResourceDisabled(o); ok {
+					t.Errorf("%d: expected resource to be enabled : %v", i, o)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:318]:
```suggestion
					t.Errorf("IsResourceDisabled(%v) = true, want false", o)
```

File: pkg/utils/apiresources_test.go [hunk 13]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+				t.Fatalf("Parse() returned error: %v", err)
 			}
-		}
-		for i, o := range test.enabled {
-			ok := r.IsResourceDisabled(o)
-			if ok {
-				t.Errorf("%d: unexpected error: %v", i, o)
+			for i, o := range test.disabled {
+				if ok := r.IsResourceDisabled(o); !ok {
+					t.Errorf("%d: expected resource to be disabled : %v", i, o)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:387]:
```suggestion
					t.Errorf("IsResourceDisabled(%v) = false, want true", o)
```

File: pkg/utils/apiresources_test.go [hunk 14]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+				}
 			}
-		}
+
+			for i, o := range test.enabled {
+				if ok := r.IsResourceDisabled(o); ok {
+					t.Errorf("%d: expected resource to be enabled : %v", i, o)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:393]:
```suggestion
					t.Errorf("IsResourceDisabled(%v) = true, want false", o)
```

File: pkg/utils/apiresources_test.go [hunk 15]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+		})
 	}
 }
 
-func TestDefaultDisabledResourceConfigGroupVersionKindParse(t *testing.T) {
+func TestResourceConfigIsEmpty(t *testing.T) {
 	tests := []struct {
-		disabled []schema.GroupVersionKind
-		enabled  []schema.GroupVersionKind
+		input string
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:402]: could you please add a test name?

`In most tests, your test failure messages should include the function inputs if they are short. If the relevant properties of the inputs are not obvious (for example, because the inputs are large or opaque), you should name your test cases with a description of what’s being tested, and print the description as part of your error message.`

https://go.dev/wiki/TestComments#identify-the-input

File: pkg/utils/apiresources_test.go [hunk 16]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
+			input: "v1/Node,Pod;networking.k8s.io;apps/v1;authorization.k8s.io/v1/SelfSubjectRulesReview",
+			want:  false,
 		},
 	}
 	for _, test := range tests {
-		r := NewDisabledResourceConfig()
-		if err := r.Parse(""); err != nil {
+		r := NewResourceConfig(true)
+		if err := r.Parse(test.input); err != nil {
 			t.Fatalf("Unexpected error: %v", err)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:417]: t.Fatalf("Parse() returned error: %v", err)

File: pkg/utils/apiresources_test.go [hunk 17]
```diff
@@ -228,144 +228,196 @@ func TestResourceConfigGroupParse(t *testing.T) {
...
-				t.Errorf("%d: unexpected error: %v", i, o)
-			}
-		}
-		for i, o := range test.enabled {
-			ok := r.IsResourceDisabled(o)
-			if ok {
-				t.Errorf("%d: unexpected error: %v", i, o)
-			}
+		if got := r.IsEmpty(); got != test.want {
+			t.Errorf("Unexpected result: %v", got)
```
Comment (zhiying-lin) [inline pkg/utils/apiresources_test.go:420]:
```suggestion
			t.Errorf("IsEmpty() = %v, want %v", got, test.want)
```

File: pkg/utils/apiresources_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/utils/apiresources_test.go]: LGTM

File: pkg/utils/apiresources_test.go (hunk 11 above)
Comment (vasudev-bongale) [reply pkg/utils/apiresources_test.go]: I ended up refactoring all the tests in apiresources_test.go, I have added small utility functions to make tests more readable and not repeating code multiple times. please let me know if that looks good.

File: pkg/controllers/clusterresourceplacement/resource_selector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasudev-bongale) [reply pkg/controllers/clusterresourceplacement/resource_selector.go]: Thanks for the explanation! I have kept the original nil check since `r.ResourceConfig.IsResourceDisabled(gvk) should return correctly if the list is empty.`

File: pkg/controllers/clusterresourceplacement/resource_selector.go [hunk 18]
```diff
@@ -287,9 +287,9 @@ func (r *Reconciler) fetchAllResourcesInOneNamespace(namespaceName string, place
 	return resources, nil
 }
 
-// shouldSelectResource returns whether a resource should be propagated
+// shouldSelectResource returns whether a resource should be selected for propagation.
 func (r *Reconciler) shouldSelectResource(gvr schema.GroupVersionResource) bool {
-	if r.DisabledResourceConfig == nil {
+	if r.ResourceConfig == nil {
```
Comment (zhiying-lin) [inline pkg/controllers/clusterresourceplacement/resource_selector.go:292]:
```suggestion
       // By default, all of the APIs are allowed.
	if r.ResourceConfig == nil {
```

nit, add a comment to improve the readability

File: pkg/resourcewatcher/change_dector.go [hunk 19]
```diff
@@ -186,7 +186,7 @@ func (d *ChangeDetector) discoverResources(dynamicResourceEventHandler cache.Res
 
 // gvrDisabled returns whether GroupVersionResource is disabled.
 func (d *ChangeDetector) shouldWatchResource(gvr schema.GroupVersionResource) bool {
-	if d.DisabledResourceConfig == nil {
+	if d.ResourceConfig == nil {
```
Comment (zhiying-lin) [inline pkg/resourcewatcher/change_dector.go:189]:
```suggestion
         // By default, all of the APIs are allowed.
	if d.ResourceConfig == nil {
```

File: pkg/controllers/clusterresourceplacement/placement_controllerv1alpha1.go [hunk 20]
```diff
@@ -49,7 +49,7 @@ type Reconciler struct {
 	// It's only needed by v1beta1 APIs.
 	UncachedReader client.Reader
 
-	// ResourceConfig contains all the api resources that we won't select based on allowed or skipped propagating apis option.
+	// ResourceConfig contains all the API resources that we won't select based on allowed or skipped propagating APIs ption.
```
Comment (zhiying-lin) [inline pkg/controllers/clusterresourceplacement/placement_controllerv1alpha1.go:52]:
```suggestion
	// ResourceConfig contains all the API resources that we won't select based on allowed or skipped propagating APIs option.
```

**Review by zhiying-lin — APPROVED** [review]
